osquery/process/processes_stats.cpp

```diff
@@ -0,0 +1,146 @@
+  if (proc_it == processes_stats_states_.end()) {
+    return std::nullopt;
+  }
+
+  return ProcessStatsSample{proc_it->second.cpu_peak_usage,
+                            proc_it->second.cpu_usages.empty()
+                                ? static_cast<std::uint16_t>(0)
+                                : proc_it->second.cpu_usages.back()};
+}
+
+void ProcessesStats::updateProcessStats(std::uint32_t pid,
+                                        std::uint64_t total_cpu_time,
+                                        std::uint32_t update_interval) {
+  std::lock_guard<std::mutex> process_stats_lock_guard(process_stats_lock_);
+  auto proc_it = processes_stats_states_.find(pid);
+
+  if (proc_it == processes_stats_states_.end()) {
+    processes_stats_states_.emplace(
+        pid, ProcessStatsState{total_cpu_time, current_generation_.load()});
+  } else {
+    auto& proc_stats = proc_it->second;
+
+    /* If this pid is older than one generation,
+       it means it's a pid that disappeared
+       before the old pid cleanup procedure had a chance to run,
+       and now has reappeared */
+    if (proc_stats.generation < current_generation_ &&
+        (current_generation_ - proc_stats.generation) > 1) {
+      // Therefore we reset its old cpu usages stats
+      proc_stats.cpu_usages.clear();
+    } else {
+      if (proc_stats.cpu_time > total_cpu_time) {
+        /* This should not normally happen but if there's pid reuse,
+           or somehow we got a bogus sample, we need to remove the old stats
+           and treat this as a new sample */
+        processes_stats_states_.erase(proc_it);
+        processes_stats_states_.emplace(
+            pid, ProcessStatsState{total_cpu_time, current_generation_.load()});
+        return;
+      }
+
+      auto delta_cpu_time = total_cpu_time - proc_stats.cpu_time;
+      auto new_cpu_usage = static_cast<uint16_t>(delta_cpu_time * 100ULL /
+                                                 (update_interval * 1000ULL));
+      proc_stats.cpu_usages.push_back(new_cpu_usage);
+
+      auto cpu_usage_max_it = std::max_element(proc_stats.cpu_usages.begin(),
+                                               proc_stats.cpu_usages.end());
+
+      proc_stats.cpu_peak_usage =
+          cpu_usage_max_it != proc_stats.cpu_usages.end() ? *cpu_usage_max_it
+                                                          : 0;
+    }
+
+    proc_stats.generation = current_generation_;
+    proc_stats.cpu_time = total_cpu_time;
+  }
+}
+
+void ProcessesStats::cleanupProcessStats() {
+  auto now = getUnixTime();
+
+  /* We base our logic for cleanup on a generation system,
+     because everytime we do a scan of the processes, we add their stats to the
+     "database"; the faster we do this, and if there are a lot of processes that
+     come and go, the more expired processes stats will be in the "database",
+     consuming resources. For this reason, every 10 generations we clean up.
```

Comment:
What's the point of waiting 10 generations? Is cleanup expensive? Is it cheaper to mark them stale?

In a heavily loaded system, I feel like there's an upper bound based on namespacesize * number of samples * sample size, and I can't tell if that's unreasonable

Reply:
I probably need to do more performance profiling to give a more direct answer.

I was worrying about having to continuously loop through several thousands of entries at each sample, to only find to cleanup few of them, so doing mainly busy work.

At the same time the 10 generations are not always followed because they depend on the sample time and if you have a lot of new processes being created and then sampled by the service, you clearly can create some buildup.

Reply:
Resolved below (https://github.com/osquery/osquery/pull/7597#issuecomment-1245827168)